Initial I2C hardware interface.

florianhumblot opened the pull request:
Tested using the display module.

Commit:
Initial i2c interface

Commit:
Need the size to be const

Commit:
Fixed the implementation.\nI was overwriting the registers instead of setting flags...

Commit:
Updated travis.yml to the build-master

Commit:
Fixed the includes according to r2d2 style

gerritvanos committed:
deleted catch part of travis

gerritvanos reviewed (approved):
since travis passed i think it is okay

wds444 reviewed (dismissed):
This code will never run, try to reduce the amount of 'dead code' in your project.

Apart from this the code seems clean, and can be merged after change.

`void i2c_bus_c::init() {
        static bool init_done = false;
        if (init_done) {
            return;
}`

florianhumblot commented:
@wds444 I'm sorry, could you please add your comment to the piece of code that will never run or 'dead code' ? that way I'll be able to see what you mean.

~~The static variable is only initialized the first time it is run, and then keeps the value. So the first run it doesn't enter the `if` and actually initializes the hardware I2C and from then on, if you call it again `init_done` will be `true` and so the function won't do anything.
This is the intended behavior and I'm not sure what you mean that it's dead code~~

My bad, I forgot to re-add `init_done = true;` when splitting the hpp into hpp and cpp...

florianhumblot committed:
fixed the static init check

Something went wrong when splitting the code to cpp and hpp files and cleaning up the functions, forgot to re-add the `init_done = true`

LRstudentHU commented on code/src/i2c_bus.cpp:
```diff
@@ -0,0 +1,148 @@
+        _selected->TWI_CR |= TWI_CR_MSDIS;
+        _selected->TWI_CR |= TWI_CR_MSEN;
+
+        static constexpr uint32_t masterClock =
```
Why is this static?

florianhumblot replied:
Why shouldn't it be? We aren't changing it afterwards

LRstudentHU replied:
Then only `constexpr` will suffice. It's very minor, but the compiler isn't allowed to omit `static constexpr` variables but it can omit `constexpr`. So technically this could prevent an optimization.

LRstudentHU commented on code/src/i2c_bus.cpp:
```diff
@@ -0,0 +1,148 @@
+        if (init_done) {
+            return;
+        }
+        if (_selected == nullptr) {
```
Can we think of a way that this check doesn't need to happen?
Maybe the selected interface is a template variable (see llib), so the code won't compile without a selected TWI.
I don't think a silent return on a nullptr selected is a good way to error out.

florianhumblot replied:
I'm unsure, if you have a good way of doing this, please go ahead and submit a patch

LRstudentHU replied:
Maybe a assert or HWLIB_FATAL?

wds444 committed:
Added explanation to what i meant in review

I used the github web editor to change the file, so i did not want to make any code changes without having the ability to test it, instead i placed comments in the code at the place i meant in my innitial request for changes.

LRstudentHU commented on code/src/i2c_bus.cpp:
```diff
@@ -0,0 +1,148 @@
+    void i2c_bus_c::write(const uint_fast8_t address, const uint8_t data[],
+                          const size_t &n) {
+        _selected->TWI_MMR = 0; ///< Reset master mode register
+        _selected->TWI_MMR = 0 << 12 | address << 16; ///< Set write and address
```
Do you need to mask the address with 0x7F, since the address is 7 bits?

florianhumblot replied:
No, we just ignore that last bit

LRstudentHU commented on code/src/i2c_bus.cpp:
```diff
@@ -0,0 +1,148 @@
+        init();
+    }
+    void i2c_bus_c::write(const uint_fast8_t address, const uint8_t data[],
+                          const size_t &n) {
```
You don't need to pass a size_t by ref

florianhumblot replied:
See previous comment to the "you don't have to pass a uint32_t by ref"

LRstudentHU commented on code/src/i2c_bus.cpp:
```diff
@@ -0,0 +1,148 @@
+        return _selected->TWI_RHR;
+    }
+    i2c_bus_c::i2c_bus_c(const interface &selected_interface,
+                         const uint32_t &SPEED)
```
You don't need to pass a uint32_t by ref

florianhumblot replied:
Const ref is a good default if you aren't going to modify the value imo. Do you have a reason why it shouldn't be done this way?

LRstudentHU replied:
You use references for two reasons:
 - Copies are too expensive
 - You are interested in changing the value of the var at the call site

If the ref is const, only the first point applies. These values fit into a single register and as such don't require a reference. 
Additional reading:

https://stackoverflow.com/questions/270408/is-it-better-in-c-to-pass-by-value-or-pass-by-constant-reference

LRstudentHU commented on code/src/i2c_bus.cpp:
```diff
@@ -0,0 +1,148 @@
+                         const uint32_t &n) {
+
+        _selected->TWI_MMR = 0; ///< Reset master mode register
+        _selected->TWI_MMR = 1 << 12 | address << 16; ///< Set read and address
```
Mask, see above

florianhumblot replied:
See above

LRstudentHU commented on code/src/i2c_bus.cpp:
```diff
@@ -0,0 +1,148 @@
+        };
+    }
+    void i2c_bus_c::read(const uint8_t address, uint8_t *data,
+                         const uint32_t &n) {
```
You don't need to pass uint32_t by ref

florianhumblot replied:
See previous

florianhumblot commented:
@wds444 your comment has been addressed and fixed

Commit:
Addressed the lead's comments on passing by reference and added an assert to hopefully crash the program if you somehow get to the initialization without having selected an interface

Commit:
Merge branch 'feature-I2C_interface' of github.com:r2d2-2019/I2C_library into feature-I2C_interface

Commit:
Removed the unnecessary comment that was incorrect

florianhumblot commented:
@LRstudentHU Your comments have been addressed, the only thing I'm unsure of is the assert in the init function. 
I wasn't able to find a hwlib_fatal which you spoke of, so if you can point me in a different direction, please do.

@wds444 your comment was correct, but it has been fixed. I forgot to re-add the line of code setting the boolean to true, but it works now and it isn't  'dead code'

Commit:
Doxygen

Commit:
Added a makefile.inc for including this library.

Commit:
Added a readme with instructions

itzandroidtab commented on code/src/i2c_bus.cpp:
```diff
@@ -0,0 +1,147 @@
+        if (init_done) {
+            return; // making this 'dead code'
+        }
+        if (_selected == nullptr) {
```
You can use the HWLIB_PANIC_WITH_LOCATION here

LRstudentHU replied:
@florianhumblot this addresses your question above.

florianhumblot replied:
I've added the macro instead.

Commit:
Remove the assert and associated headerfile and added a HWLIB panic instead.

Commit:
Removed outdated comment
